src/transformers/lib/rehypeSqueezeBreaks.ts

```diff
@@ -0,0 +1,152 @@
+  node.type === 'text' && node.value.replace(/\s/g, '') === ''
+
+const isBr = (node: RootContent) =>
+  node.type === 'element' && node.tagName === 'br'
+
+const squeezeSoftBreaks = ({
+  children,
+  maxSoftBreaks,
+}: { children: ElementContent[] } & Pick<Props, 'maxSoftBreaks'>) => {
+  const newChildren: ElementContent[] = []
+  const isRetainAll = maxSoftBreaks === -1
+  let breakCount = 0
+
+  children.forEach((node) => {
+    if (node.type !== 'element' || node.tagName !== 'br') {
+      breakCount = 0
+      newChildren.push(node)
+      return
+    }
+
+    // cap empty paragraphs or retain all by adding <br>
+    breakCount++
+    const shouldRetain =
+      isRetainAll || (maxSoftBreaks ? breakCount <= maxSoftBreaks : false)
+    if (shouldRetain) {
+      newChildren.push({
+        type: 'element',
+        tagName: 'br',
+        properties: {},
+        children: [],
+      })
+    }
+  })
+
+  return newChildren
+}
+
+const squeezeHardBreaks = ({
+  children,
+  maxHardBreaks,
+  maxSoftBreaks,
+}: { children: Array<RootContent | ElementContent> } & Props) => {
+  const newChildren: RootContent[] = []
+  const isRetainAll = maxHardBreaks === -1
+  let breakCount = 0
+
+  children.forEach((node) => {
+    // skip empty text nodes
+    if (isEmptyText(node)) {
+      newChildren.push(node)
+      return
+    }
+
+    // paragraphs in blockquote
+    if (node.type === 'element' && node.tagName === 'blockquote') {
+      newChildren.push({
+        type: 'element',
+        tagName: 'blockquote',
+        properties: node.properties,
+        children: squeezeHardBreaks({
+          children: node.children,
+          maxHardBreaks,
+          maxSoftBreaks,
+        }) as ElementContent[],
+      })
+      return
+    }
+
+    // skip non-paragraph node
+    if (node.type !== 'element' || node.tagName !== 'p') {
+      breakCount = 0
+      newChildren.push(node)
+      return
+    }
+
+    // skip non-empty paragraph:
+    // - <p></p>
+    // - <p>  </p>
+    // - <p><br></p>
+    // - <p>  <br></p>
+    const isEmptyParagraph =
+      node.children.length === 0 ||
+      node.children.every((n) => isBr(n) || isEmptyText(n))
+    if (!isEmptyParagraph) {
+      breakCount = 0
+      newChildren.push({
+        type: 'element',
+        tagName: 'p',
+        properties: node.properties,
+        children: squeezeSoftBreaks({
+          children: node.children,
+          maxSoftBreaks,
+        }),
+      })
+      return
+    }
+
+    // cap empty paragraphs or retain all by adding <br>
+    breakCount++
+    const shouldRetain =
+      isRetainAll || (maxHardBreaks ? breakCount <= maxHardBreaks : false)
+    if (shouldRetain) {
+      newChildren.push({
+        type: 'element',
+        tagName: 'p',
+        properties: {},
+        children: [
+          {
+            type: 'element',
+            tagName: 'br',
+            properties: {},
+            children: [],
+          },
+        ],
+      })
+    }
+  })
```

Comment:
Switch case might be more readable

```suggestion
    if (node.type === 'element') {
    switch (node.tagName) {
      case 'blockquote':
        newChildren.push(createNode('blockquote', node.properties, squeezeHardBreaks({
          children: node.children,
          maxHardBreaks,
          maxSoftBreaks,
        }) as ElementContent[]));
        break;
      case 'p':
        const isEmptyParagraph = node.children.length === 0 || node.children.every((n) => isBr(n) || isEmptyText(n));
        if (!isEmptyParagraph) {
          newChildren.push(createNode('p', node.properties, squeezeSoftBreaks({
            children: node.children,
            maxSoftBreaks,
          })));
        } else {
          // cap empty paragraphs or retain all by adding <br>
          breakCount++;
          const shouldRetain = isRetainAll || (maxHardBreaks ? breakCount <= maxHardBreaks : false);
          if (shouldRetain) {
            newChildren.push(createEmptyNode('p'));
          }
        }
        break;
      default:
        breakCount = 0;
        newChildren.push(node);
        break;
    }
  }
```

Reply:
good idea, resolved